DM-2961: Port the psfextractor external library from HSC to LSST

mjuric opened the pull request:
Creating a PR to make it easier to leave comments

natelust committed:
Introduce code to replace dependency on clapack

This code re-implements functionality that is normally provided through a
lapack implementation. It was undesirable to include an additional
dependency in the LSST stack, and because of the poor structure of the
psfex codebase it was easier to create a library that pretended to be a
full lapack implementation and ship it with psfex. Because it is not a full
implementation (it only implements two functions) it was best not to ship
it as an additional LSST package, so no future developer would mistake the
functionality.

natelust committed:
Script to control the eupspkg build process

natelust committed:
Introduced scons built files

natelust committed:
Removed lapack dependency, added library export

The package dependency (from the hsc source) on clapack was removed.
Additionally the path to the dynamic library for the clapack replacement is
added such that the meas_exetensions_psfex package can locate it.

Add gsl as a required in ups table

natelust committed:
Remove unneeded eups hsc build file

mjuric commented on hack_clapack/.sconf_temp/build.cfg:
```diff
@@ -0,0 +1,4 @@
+[Build]
+cc = clang
+opt = 0
+
```
I think this file (and all others in .scons_temp) shouldn't be in the git repo (they're auto generated).

timj replied:
I was wondering about `sconf_temp` vs `scons_temp`

mjuric replied:
sorry, should've said `.sconf_temp` -- it's auto generated as well.

timj replied:
I will just state in passing the evils of `git commit -a`... (just a guess here :smile:)

mjuric commented:
`hack_clapack/src/clapack.os` looks like a build product -- shouldn't be in the git repo. I think it'd be good to scrub the repository for these -- many seem to have snuck in.

mjuric commented on ups/eupspkg.cfg.sh:
```diff
@@ -0,0 +1,29 @@
+	options=''
+	options=$options"--prefix=$PREFIX "
+	options=$options"--disable-threads "
+	options=$options"--with-clapack=$PSFEX_DIR/hack_clapack "
```
It will be more readable (consistent with other `eupspkg.cfg.sh` examples) it you place all of these on a single line, i.e.:

```
options="--prefix=$PREFIX --disable-threads --with-clapack=$PSFEX_DIR/hack_clapack"
```

natelust replied:
done

mjuric commented on ups/eupspkg.cfg.sh:
```diff
@@ -0,0 +1,29 @@
+	if [ "$FFTW_DIR" ]; then
+		options=$options"--with-fftw-incdir=$FFTW_DIR/include "
+		options=$options"--with-fftw-libdir=$FFTW_DIR/lib "
+	fi
```
Same as above; you could do:

```
options+=" --with-fftw-incdir=$FFTW_DIR/include --with-fftw-libdir=$FFTW_DIR/lib"
```

mjuric commented on ups/eupspkg.cfg.sh:
```diff
@@ -0,0 +1,29 @@
+	options=$options"--prefix=$PREFIX "
+	options=$options"--disable-threads "
+	options=$options"--with-clapack=$PSFEX_DIR/hack_clapack "
+	if [ "$FFTW_DIR" ]; then
```
It may be better to explicitly test if the variable is empty:

```
if [[ -z "$FFTW_DIR" ]]; then
```

mjuric replied:
Actually, you don't need this line at all, since you have `setupRequired(fftw)` in the `.table` file, which guarantees that `FFTW_DIR` will be set. In that case, you can compress _all_ these options lines into a single one.

natelust replied:
I actually added the check for FFTW_DIR because there was a request to be able to use the system fftw. If the system fftw is used, does FFTW_DIR still get set in whatever dummy package eups provides?

mjuric replied:
Ah, so if `fftw` is declared with something like `manifest.remap`, and w/o its `.table` file? OK, I see.

In that case, add a comment above that `if` statement explaining why it's there. In general, whenever there's something in the code which may seem unusual to someone who does not know all your requirements/assumptions, document it.

PS: I kid you not -- someone just now ended a talk with: "Always code as if the guy who ends up maintaining your code will be a violent psychopath who knows where you are." :).

natelust replied:
Added comment for clarity, but left the config options within the if block

mjuric commented on ups/eupspkg.cfg.sh:
```diff
@@ -0,0 +1,29 @@
+		options=$options"--with-fftw-libdir=$FFTW_DIR/lib "
+	fi
+	#echo $options
+	pwd
```
You don't need the above two lines.

natelust replied:
removed

mjuric commented on ups/eupspkg.cfg.sh:
```diff
@@ -0,0 +1,29 @@
+}
+
+build(){
+	make CFLAGS=-g
```
If you need to override CFLAGS explicitly, add to them rather than overwriting:

```
make CFLAGS="$CFLAGS -g"
```

natelust replied:
That was left over from hunting a dangling pointer bug, removed now

mjuric commented on ups/eupspkg.cfg.sh:
```diff
@@ -0,0 +1,29 @@
+	scons opt=3 prefix=$PREFIX version=$VERSION
+}
+
+install(){
```
Start this function by invoking `clean_old_install`. E.g., see https://github.com/lsst/boost/blob/master/ups/eupspkg.cfg.sh for an example.

natelust replied:
added

mjuric commented on ups/eupspkg.cfg.sh:
```diff
@@ -0,0 +1,29 @@
+
+install(){
+	make install
+	scons opt=3 install prefix=$PREFIX version=$VERSION
```
You don't want to be explicitly setting `opt=3` here -- that will already be (automatically) set through `SCONSFLAGS`.

natelust replied:
removed in both places

timj commented on hack_clapack/Sconstruct:
```diff
@@ -0,0 +1,12 @@
+source = ["src/clapack.c"]
+libs   = ['gsl','gslcblas']
+lib_path = os.environ['LD_LIBRARY_PATH'].split(':')
+env['CPPPATH']=[os.environ['GSL_DIR']+'/include','/usr/include','/usr/local/include']
```
I'd prefer `os.path.join` here rather than just appending `/include`

timj commented:
One minor comment on python coding style. Can you be a bit more closer to PEP-8 by leaving spaces after commas and the like? I know we haven't yet defined our coding standard in terms of PEP-8 but a quick run of `flake8` would be helpful.

mjuric commented:
@natelust I did a quick review of the eups packaging parts -- I think here's enough comments in there for you to digest/fix up for now.

mjuric commented on ups/eupspkg.cfg.sh:
```diff
@@ -0,0 +1,29 @@
+
+config(){
```
Another minor nit -- though we don't have an official `bash` coding standard, all of the eupspkg files use a convention where the opening brace `{` is on the line after the function name definition. E.g., https://github.com/lsst/boost/blob/master/ups/eupspkg.cfg.sh.

PS: Tabs are good!

natelust committed:
iChanged name of hack_clapack

The name is now lapack_functions to make it more clear we are only
implementing some of the functions from lapack

natelust committed:
Clean up for clarity and functionality

Added clean_old_install to the install function. Formatting with braces to
come after the function declaration

The options were condensed down to one line. Comments were added to explain
why the fftw configuration is optional. Removed unneeded lines of code
which had been used for debug

natelust commented:
Fixed Sconstruct issues, and tested with flake8. The two remaining flake8 issues are either related to this being a scons file (undefined name 'Environment') or a LSST standard (line too long (103 > 79 characters) where we allow 110)

natelust committed:
Removed test file used in development

This file was used to test that the functionality reproduced that of
clapack, but is now unneeded

natelust committed:
Fixed path being incorrectly joined

Updated lapack_functions Sconstruct to pep8

natelust committed:
Comments added and non-needed code removed

natelust committed:
Pointed to updated fftw package

Single and double preseason libraries are now exported with the same
description

natelust committed:
More robust setting of lapack path

natelust committed:
Scons path for lapack_functions

natelust committed:
Clean up of build files

Streamlined the process of building with scons. Explicitly copied the
lapack_functions library into the psfex/libs dir. This allows everything
building and using this library to find it more effectively.

natelust committed:
Remove unneeded references to blas from configure

These only served as a check in the origional psfex code, and are not used
in our version. These lines were causing some tests to fail

natelust committed:
Introduce dummy blas library

Like the f2c library, this is introduced simply because the build system
expects to see it, but we no longer use the code. It is simply a place
holder.

jdswinbank commented:
This work is long since complete.
